### PR TITLE
Add Objection w/ initial schema and seeds, and auth w/ crypto

### PR DESCRIPTION
**Pull request by samhh**
This PR includes database connectivity, initial schema, initial seeds, and the automatic creation of a new player/account on server start as a demo.

Included also is MJS so that we can use modern JS import/export syntax isomorphically.

### CODE REVIEW

File: server/db/migrations/20171020144702_initial_schema.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+			table.dateTime('last_active').defaultTo(knex.fn.now()).notNullable()
+			table.string('username', 20).unique().notNullable()
+			table.dateTime('last_modified_username')
+			table.string('email', 254).unique()
+			table.string('password', 100).notNullable() // TODO update this length to reflect pw crypto
+			table.integer('selected_flair').references('id').inTable('flair').onDelete('SET NULL')
+			table.integer('selected_ball').references('id').inTable('ball').onDelete('SET NULL')
+			table.specificType('competitive_honors', 'text[]')
+		})
+		.createTable('map', table => {
```
Comment (RonSpawnson) [inline server/db/migrations/20171020144702_initial_schema.js]: Perhaps maps should have author information. Also note, some maps have more than 1 author

Comment (samhh) [reply server/db/migrations/20171020144702_initial_schema.js]: Good shout, I'll add an authors key as something like `varchar(n)[]`.

Worth adding the same to flairs and balls or no?

File: server/db/migrations/20171020144702_initial_schema.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
+exports.up = knex =>
+	knex.schema
+		.raw('create extension if not exists "uuid-ossp"')
+		.createTable('flair', table => {
+			table.increments('id').primary().notNullable()
+			table.string('name', 20).notNullable()
+			table.string('description', 100)
+			table.text('filename').notNullable()
+		})
+		.createTable('ball', table => {
```
Comment (whackashoe) [inline server/db/migrations/20171020144702_initial_schema.js]: what is the point of this table? for custom balls?

File: server/db/migrations/20171020144702_initial_schema.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+			table.increments('id').primary().notNullable()
+			table.string('name', 20).notNullable()
+			table.string('description', 100)
+			table.text('filename').notNullable()
+		})
+		.createTable('player', table => {
```
Comment (whackashoe) [inline server/db/migrations/20171020144702_initial_schema.js]: wins, losses?

File: server/db/migrations/20171020144702_initial_schema.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+exports.up = knex =>
```
Comment (whackashoe) [inline server/db/migrations/20171020144702_initial_schema.js:1]: would be cool to have games table which stores all existing game stats,

players
joined_at

who scored
who won
etc

File: server/db/migrations/20171020144702_initial_schema.js (hunk 2 above)
Comment (samhh) [reply server/db/migrations/20171020144702_initial_schema.js]: Yep, that's the idea. That can be removed for the time being though as it isn't fleshed out and certainly won't be in MVP.

File: server/db/migrations/20171020144702_initial_schema.js (hunk 3 above)
Comment (samhh) [reply server/db/migrations/20171020144702_initial_schema.js]: Yes... this needs discussion I think about what stats we're storing and in what format.

File: server/db/migrations/20171020144702_initial_schema.js (hunk 4 above)
Comment (samhh) [reply server/db/migrations/20171020144702_initial_schema.js:1]: Would this be for temporary use? If so, probably worth consulting with @OpenSourceCTF/engine requirements.

File: server/db/migrations/20171020144702_initial_schema.js (hunk 1 above)
Comment (sallaben) [reply server/db/migrations/20171020144702_initial_schema.js]: Since when do flairs and balls have authors? Or are you implying something else?

Comment (samhh) [reply server/db/migrations/20171020144702_initial_schema.js]: I'm imagining a system whereby flairs/balls/other assets are contributed to the game by the community. In that case they may like accreditation as map authors do.

**Comment by sallaben** [comment]
That's a good idea, actually. Maybe we could create an in-browser flair
creator. That would put a spin on things.

File: server/db/migrations/20171020144702_initial_schema.js (hunk 1 above)
Comment (SamHH) [reply server/db/migrations/20171020144702_initial_schema.js]: How do we want authors to be denoted? I was thinking an array of strings, but then I wondered if we might prefer actually linking them to the players i.e. an array of player IDs (corresponding to the player table) instead.

**Comment by samhh** [comment]
I've removed the balls and flairs tables as we probably won't be touching this for MVP (and if we are, we can easily add them back in with better requirements). Same goes for competitive honours column.

I've added an authors column to the map table as an array of strings.

What's left for this PR:

- Determine what stats we want to store, how we want to store them, and whether that's something for MVP or not.

**Comment by sallaben** [comment]
When we do implement it, I think stats should be stored by the engine
during each game and committed to the database whenever the user
disconnects from that game (whether through quitting or joining a new one
when that game ends). What do you think?

**Comment by samhh** [comment]
Note to self: All that's left before this branch can be merged is actioning Contessa's feedback in Discord/#backend.

**Comment by samhh** [comment]
RE: the above, that's all that needs doing now. The other advice was regards matches which can come later. Time this PR was merged!

**Pull request merged**